sympy/polys/rings.py

```diff
@@ -3005,3 +3027,243 @@ def sqf_list(f, all=False):
+    >>> R, x, y = ring("x, y", ZZ)
+
+    >>> f = x + y
+    >>> g = g = x**2*y + x*y
+    >>> p = [f, g]
+    >>> monomial_extract(p)
+    ([x + y, x**2*y + x*y], None)
+
+    >>> f = x**2*y
+    >>> g = x**2*y + x*y
+    >>> p = [f, g]
+    >>> monomial_extract(p)
+    ([x**2*y, x**2*y + x*y], x*y)
+
+    """
+    ring = p[0].ring
+    zero_monom = ring.zero_monom
+
+    monomials = chain(*p)
+    monom_gcd = tuple(map(min, zip(*monomials)))
+
+    # If the zero monomial is present, we can immediately return
+    if monom_gcd == zero_monom:
+        return p, None
+    else:
+        # Loop over a few monomials to quickly check if gcd is zero
+        for i, monom in enumerate(monomials):
+            if i >= 2:
+                break
+            gcd_monom = tuple(map(min, zip(monom_gcd, monom)))
+            if gcd_monom == zero_monom:
+                return p, None
+
+        d = ring({monom_gcd: ring.domain.one})
+        _d = degree_list(d.as_expr())
+        print(_d)
+        p_result = []
+        for pi in p:
+            pi = degree_list(pi.as_expr())
+            print(pi)
+            p_result.append(ring.monomial_ldiv(pi, _d))
+        p_result = p
+        return p, d
+
+def gcd_ground(coeffs, domain):
+    """
+    Compute the greatest common divisor (GCD) of elements in the ground domain.
+
+    Examples
+    ========
+
+    >>> from sympy import ZZ
+    >>> coeffs = [3, 12, 6]
+    >>> gcd_ground(coeffs, ZZ)
+    3
+
+    """
+    coeffs = list(coeffs)
+    gcd = domain.gcd
+    d = coeffs[0]
+
+    for coeff in coeffs[1:]:
+        d = gcd(d, coeff)
+
+        if d == domain.one:
+            break
+
+    return d
+
+def gcd_coeffs(p):
+    """
+    Simplify a list of polynomials whose gcd is wanted. Returns a possibly
+    longer list of simpler polynomials having the same gcd as the input.
+
+    This is done by eliminating symbols that can not be part of the gcd
+    because they do not appear in each item of the input. In the output
+    list, all items have exactly the same symbols. The set of those symbols
+    is also returned.
+
+    Examples
+    ========
+
+    >>> from sympy import ring, ZZ
+    >>> R, x, y = ring("x, y", ZZ)
+
+    >>> f = x**2 - y**2
+    >>> g = x - y
+    >>> p = [f, g]
+    >>> gcd_coeffs(p)
+    ([1], None)
+
+    """
+    all_coeffs = p
+
+    while True:
+        # Quick exits are most efficient if we start from the simplest polys
+        p = sorted(set(all_coeffs), key=len)
+
+        # Find the intersection of symbols for each poly:
+        common = p[0].main_variable()
+        allsame = True
+        for pi in p[1:]:
+
+            # Quick exit
+            if not common:
+                R = p[0].ring
+                K = R.domain
+                gcd = gcd_terms(p, R, K)
+                return [gcd], None
+
+            syms = pi.main_variable()
+            if allsame and syms != common:
+                allsame = False
+            common &= syms
```

Comment:
Here `main_variable` only returns a single generator even when the poly has multiple. Maybe I misunderstood but just checking that all polys have the same main variable is not the same as checking that they all have the same symbols.

Please add some tests that explore different cases.

Reply:
It will return the smallest index among all generators.
for example:
`p = x*y**2 - y**3 + 4*y` 
output is `0`

Reply:
This was marked as resolved when it should not have been.

Again saying `syms = pi.main_variable` is confusing. You are assigning a singular object "variable" to a plural name "syms".

Also the logic here is not correct: the intersection of the set of main variables is not the same as the intersection of the set of *all* variables present.

Reply:
Is this logic correct?
comparing the `main_variable` of each polynomial with the current value of common
``` python
for pi in polynomials:
    main_sym = pi.main_variable()

    if allsame and main_sym != common:
        allsame = False
    common &= main_sym 
```